app/services/declarations/void.rb

```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+module Declarations
+  class Void
+    include ActiveModel::Model
+    include ActiveModel::Attributes
+
+    attribute :declaration
+
+    validates :declaration, presence: true
+
+    validate :declaration_not_already_voided, if: :voiding?
+    validate :declaration_is_voidable, if: :voiding?
+
+    validate :declaration_not_already_refunded, if: :clawing_back?
+    validate :output_fee_statement_available, if: :clawing_back?
+
+    def void
+      return false unless valid?
+
+      ApplicationRecord.transaction do
+        clawing_back? ? clawback_declaration : void_declaration
+      end
+
+      true
+    end
+
+  private
+
+    def clawback_declaration
+      declaration.awaiting_clawback_state!
+      statement_attacher.attach
+    end
+
+    def void_declaration
+      declaration.voided_state!
+      declaration.statement_items.with_state(:eligible, :payable).first&.mark_voided!
+    end
+
+    def declaration_not_already_voided
+      return unless declaration
+
+      errors.add(:declaration, :already_voided) if declaration.voided_state?
+    end
+
+    def declaration_is_voidable
+      return unless declaration
+
+      errors.add(:declaration, :not_voidable) unless declaration.voidable?
+    end
+
+    def declaration_not_already_refunded
+      return unless declaration
+
+      already_refunded = declaration.statement_items.refundable.exists?
+      errors.add(:declaration, :not_already_refunded) if already_refunded
+    end
+
+    def output_fee_statement_available
+      errors.add(:declaration, :no_output_fee_statement, cohort: declaration.cohort) unless statement_attacher.valid?
+    end
+
+    def statement_attacher
+      @statement_attacher ||= StatementAttacher.new(declaration:)
+    end
+
+    def clawing_back?
+      declaration&.paid_state?
```

Comment:
There is an issue in the old implementation that if someone hits the void endpoint for a paid declaration, which then moved to awaiting clawback, and hits it again we get the voidable error which is confusing. Maybe we can include paid/awaiting clawback/clawed back states here to ensure it goes down the right path and returns the right error, happy to discuss and explain further

Reply:
I think I follow; I've used the error we had for this scenario in ECF (although the bug meant it was not reachable in ECF). [Here's the change](https://github.com/DFE-Digital/npq-registration/pull/1460/commits/673f811f05ccce26ba428b0cc9b9389382f415ec).